From e26c6e832691626281c4be0af270e6e139bf40e6 Mon Sep 17 00:00:00 2001
From: Will Dietz <w@wdtz.org>
Date: Tue, 3 Jan 2017 08:03:30 -0600
Subject: [PATCH 8/8] kludge around _DYNAMIC and our bad weak symbol handling.

---
 src/env/__init_tls.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/env/__init_tls.c b/src/env/__init_tls.c
index b125eb1f..f9d56742 100644
--- a/src/env/__init_tls.c
+++ b/src/env/__init_tls.c
@@ -71,8 +71,8 @@ typedef Elf32_Phdr Phdr;
 typedef Elf64_Phdr Phdr;
 #endif
 
-__attribute__((__weak__, __visibility__("hidden")))
-extern const size_t _DYNAMIC[];
+//__attribute__((__weak__, __visibility__("hidden")))
+//extern const size_t _DYNAMIC[];
 
 static void static_init_tls(size_t *aux)
 {
@@ -86,8 +86,10 @@ static void static_init_tls(size_t *aux)
 		phdr = (void *)p;
 		if (phdr->p_type == PT_PHDR)
 			base = aux[AT_PHDR] - phdr->p_vaddr;
+		/*
 		if (phdr->p_type == PT_DYNAMIC && _DYNAMIC)
 			base = (size_t)_DYNAMIC - phdr->p_vaddr;
+		*/
 		if (phdr->p_type == PT_TLS)
 			tls_phdr = phdr;
 	}
-- 
2.11.0

